app/cni/cmd/install/main.go

```diff
@@ -0,0 +1,379 @@
+		if err != nil {
+			return err
+		}
+
+		marshaled, err := transformJsonConfig(kumaCniConfig, hostCniConfig)
+		if err != nil {
+			return err
+		}
+		log.V(1).Info("resulting config", "config", string(marshaled))
+
+		err = atomic.WriteFile(mountedCniNetDir+"/"+resolvedName, bytes.NewReader(marshaled))
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+	return nil
+}
+
+func fileExists(path string) bool {
+	_, err := os.Stat(path)
+	return err == nil
+}
+
+func prepareKubeconfig(mountedCniNetDir, kubeconfigName, serviceAccountPath string) error {
+	serviceAccountTokenPath := serviceAccountPath + "/token"
+	serviceAccountToken, err := ioutil.ReadFile(serviceAccountTokenPath)
+	if err != nil {
+		return err
+	}
+
+	if fileExists(serviceAccountTokenPath) {
+		kubernetesServiceHost := env.GetString("KUBERNETES_SERVICE_HOST", "")
+		if kubernetesServiceHost == "" {
+			return errors.New("KUBERNETES_SERVICE_HOST env variable not set")
+		}
+
+		kubernetesServicePort := env.GetString("KUBERNETES_SERVICE_PORT", "")
+		if kubernetesServicePort == "" {
+			return errors.New("KUBERNETES_SERVICE_PORT env variable not set")
+		}
+
+		kubeCaFile := env.GetString("KUBE_CA_FILE", serviceAccountPath+"/ca.crt")
+		kubeCa, err := ioutil.ReadFile(kubeCaFile)
+		if err != nil {
+			return err
+		}
+		kubernetesServiceProtocol := env.GetString("KUBERNETES_SERVICE_PROTOCOL", "https")
+		caData := base64.StdEncoding.EncodeToString(kubeCa)
+
+		kubeconfig := kubeconfigTemplate(kubernetesServiceProtocol, kubernetesServiceHost, kubernetesServicePort, string(serviceAccountToken), caData)
+		err = atomic.WriteFile(mountedCniNetDir+"/"+kubeconfigName, strings.NewReader(kubeconfig))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func kubeconfigTemplate(protocol, host, port, token, caData string) string {
+	safeHost := host
+	if govalidator.IsIPv6(host) {
+		if !(strings.HasPrefix(host, "[") && strings.HasSuffix(host, "]")) {
+			safeHost = "[" + host + "]"
+		}
+	}
+
+	serverUrl := url.URL{
+		Scheme: protocol,
+		Host:   safeHost + ":" + port,
+	}
+
+	return `# Kubeconfig file for kuma CNI plugin.
+apiVersion: v1
+kind: Config
+clusters:
+- name: local
+  cluster:
+    server: ` + serverUrl.String() + `
+    certificate-authority-data: ` + caData + `
+users:
+- name: kuma-cni
+  user:
+    token: ` + token + `
+contexts:
+- name: kuma-cni-context
+  context:
+    cluster: local
+    user: kuma-cni
+current-context: kuma-cni-context`
+}
+
+func copyBinaries() error {
+	dirs := []string{primaryBinDir, secondaryBinDir}
+	writtenOnce := false
+	var err error
+	for _, dir := range dirs {
+		if !isDirWriteable(dir) {
+			log.Info("directory is not writeable", "dir", dir)
+			continue
+		}
+		file, err := os.Open(kumaCniBinaryPath)
+		if err != nil {
+			log.Error(err, "can't open kuma-cni file")
+			continue
+		}
+
+		stat, err := os.Stat(kumaCniBinaryPath)
+		if err != nil {
+			log.Error(err, "can't stat kuma-cni file")
```

Comment:
yeah, we should - but I'm also thinking if we actually need the `stat/chmod` thing in the first place. I initially added that because I was getting "file not found" errors, but it turned out the real problem was "you wrongly cross-compiled this binary and the interpreter is missing". The chmod thing was not present in the previous version. I'll remove it and move the contents of the loop to a separate function.